### PR TITLE
Docker support

### PR DESCRIPTION
**Pull request by viewpnt1**
Hello,

I added an easy to use docker configuration to the project. More on how to use it is in Readme.md.

Since I have very low experience with Python and the repo itself had little documentation on how to deploy it, I tried to put together parts from dokku repo and this repo to make this as good as possible. I would recommend to inspect and apply fixes.

Along with docker I fixed a few bugs that were causing exceptions.

I hope you find this as a valuable addition to the repo, and I hope this is something that can bring it back to life.

Best regards,
Milan

### CODE REVIEW

File: gengine/app/leaderboard.py [hunk 1]
```diff
@@ -4,6 +4,7 @@
 from gengine.app.model import t_subjects, t_subjectrelations, t_subjects_subjects, Subject
 from gengine.metadata import DBSession
 
+import logging
```
Comment (feldi-online) [inline gengine/app/leaderboard.py]: __[Flake8]__ 'logging' imported but unused ([see on Sider](https://sider.review/gh/repos/31541379/pulls/46#issue-4893294))

<details>

| Rule |
| - |
| F401 |

</details>
<sub>posted by Sider</sub>

File: gengine/app/leaderboard.py [hunk 2]
```diff
@@ -13,28 +14,29 @@ class GlobalLeaderBoardSubjectSet:
...
-                #or_(
-                #    t_subjects.c.deleted_at == None,
-                #    t_subjects.c.deleted_at >= to_date
-                #)
-            ))
-        else:
-            q = q.where(or_(
-                and_(
-                    t_subjects.c.created_at <= from_date,
+        if from_date != None and to_date != None:
```
Comment (feldi-online) [inline gengine/app/leaderboard.py]: __[Flake8]__ comparison to None should be 'if cond is not None:' ([see on Sider](https://sider.review/gh/repos/31541379/pulls/46#issue-4893295))

<details>

| Rule |
| - |
| E711 |

</details>
<sub>posted by Sider</sub>

**Comment by msander** [comment]
@viewpnt1 
Hello Milan,

thank you for your pull request. Most of the changes look fine, I have just a few points to discuss:

- If we are sticking with alpine I'd like to pin the version to 3.10. I had a bad experience with alpine making incompatible changes during updates. Otherwise we could also use buster.
- How are you using the custom_tasks folder? If you want to include custom code in your application, you can use the pyramid.includes setting in the ini file. 
```
pyramid.includes =
    pyramid_debugtoolbar
    pyramid_tm
    mygengine_ext
```

With mygengine_ext as

```
def includeme(config):
    init_extension()

def init_extension():
    from gengine.app.registries import get_admin_extension_registry
    get_admin_extension_registry().register("mygengine_ext.admin_extension", admin_extension)

    from gengine.app.registries import get_task_registration_points_registry
    get_task_registration_points_registry().register("mygengine_ext.task_registration_point", task_registration_point)

def admin_extension(adminapp, admin):
   pass

def task_registration_point():
   pass
```

- In the production nginx config, you have
```
location = / {
    rewrite ^/(.*)$ /admin/$1 last;
}
```
I think this is redirecting all requests to /admin/, which will probably break the /api calls, right?

- If you are going to contribute more, I'd love to talk to you about your use-case and the project roadmap (we are also planning to integrate some new features next year)

Regards
Marcel

**Comment by viewpnt1** [comment]
Hello Marcel,

Thanks for the response. To answer your questions:

1. Alpine version pinning - I totally agree. Pining versions is always a safer approach.
2. Custom tasks might not be necessary. I only added it so I can add my own @EngineTask definitions somewhere, and keep demo task and its respective files unchanged. By doing so I believed the concept will be least intrusive to the existing codebase. Although, I would prefer your solution if it doesn't require `config.include('gengine.custom_tasks')` in gengine/__init__.py.
3. This route will redirect only exact location "/" to admin, although now that you mention it, it is kinda weird. I got the whole config made out of [your dokku config](https://github.com/ActiDoo/gamification-engine-dokku/blob/master/config/nginx.conf.erb), but haven't given it much thought.

4. My project roadmap:
I work on a web app that along core features offers gamification concept to the users. So I found your repo and decided to give it a try. Along the way I adjusted/tweaked/fixed not-so-small number of features I thought weren't working as intended (but I'm actually not sure for all of them) - for example route add_or_update_subject didn't really offer possibility to add subject, only update - stuff like that. Also, there might be some opinionated fixes I applied that were necessary for my project's use cases, so that would need to be discussed and reviewed or discarded. I would be happy to contribute parts that are suitable. 
The project is going live by the end of year so, if anything, I will be applying bug fixes regularly in the near future.

Best regards,
Milan

**Pull request merged**